modules/terminate-agent-hook/lambda/lambda_function.py

```diff
@@ -10,12 +10,105 @@
 import json
 import os
 import sys
 
 
+def check_runner_running(client, instance_id):
+    """
+    Checks if the runner instance is running.
+    :param client: the boto3 ec2 client
+    :param instance_id: the ID of the runner instance
+    :return: true or false, whether the runner instance is running
+    """
+    print(json.dumps({
+        "Level": "info",
+        "Message": "Looking for running runner instance..."
+    }))
+    try:
+        reservations = client.describe_instances(InstanceIds=[instance_id], Filters=[
+            {
+                "Name": "instance-state-name",
+                "Values": ["running", "pending"],
+            }
+        ]).get("Reservations")
+    except ClientError as error:
+        print(json.dumps({
+            "Level": "error",
+            "Message": "Failed to lookup runner instance"
+        }))
+        raise error
+
+    if len(reservations) > 0:
+        print(json.dumps({
+            "Level": "info",
+            "Message": "Runner instance still running"
+        }))
+        return True
+
+    print(json.dumps({
+        "Level": "info",
+        "Message": "Runner instance already terminated"
+    }))
+    return False
+
+
+def stop_runner_service(client, instance_id):
+    """
+    Stops the gitlab-runner service on the runner instance using SSM command.
+    The command may fail if the gitlab-runner service has jobs running, in
+    which case the function will error and be re-tried by SQS.
+    :param client: the boto3 SSM client
+    :param instance_id: the ID of the runner instance
+    """
+    print(json.dumps({
+        "Level": "info",
+        "Message": "Stopping gitlab-runner service..."
+    }))
+
+    try:
+        initial_response = client.send_command(DocumentName=os.environ['DOCUMENT_NAME'],
+            Comment="Stop gitlab-runner service, and check whether it's stopped.",
+            InstanceIds=[instance_id]
+        )
+    except ClientError as error:
+        print(json.dumps({
+            "Level": "error",
+            "Message": "Failed to send SSM command"
+        }))
+        raise error
+    
+    command_id = initial_response['Command']['CommandId']
+
+    try:
+        waiter = client.get_waiter('command_executed')
+        waiter.wait(
+            CommandId=command_id,
+            InstanceId=instance_id,
+            WaiterConfig={
+                "Delay": 3,
+                "MaxAttempts": 10
+            }
+        )
+        command_response = client.get_command_invocation(CommandId=command_id, InstanceId=instance_id)
+    except WaiterError as error:
+        print(json.dumps({
+            "Level": "error",
+            "Message": "Failure waiting for command to be successful"
+        }))
+        raise error
+
+    if command_response['Status'] == "Success":
+        print(json.dumps({
+            "Level": "info",
+            "Message": f"gitlab-runner service stopped, SSM command response: {command_response}"
```

Comment:
nitpick: This seems to be incorrect as we do not know if the service has been stopped already. It might still be running, waiting for jobs to be finished, right?

Reply:
The SSM command sent to the runner instance will stop the gitlab-runner service and only return success if the service is inactive.

Reply:
So the waiter above returns only in case the service has been stopped? This could last up to 1800s (by default), but as far as I understand the waiter, it definitely returns after 3*10 seconds, doesn't it? Is this a problem? Wouldn't we return a failure then?

Reply:
The SSM command sent to the instance does not wait for the service to finish stopping(1800s), it will:
1. attempt to stop `gitlab-runner.service`(non blocking)
2. wait 5s
3. check the status of `gitlab-runner.service`
4. exit 0(SSM command status = "Success") if the service is stopped and docker machines removed, exit 1(SSM command status = "Failed") otherwise

This `if` statement checks the status of the SSM command, not the waiter. Also, if the waiter reaches max retries and the SSM command hasn't finished running, then it should fail and the function will exit with failure to be retried via SQS.

Sorry it's a bit confusing, let me know if that doesn't make sense.